# HW8 Solution Walkthrough (Python)
#### Columbia University APAN 5310 - SQL & Relational Databases (Fall 2018)

*Prepared by Nick Machairas. Posted on Nov. 13, 2018*

---

First, import necessary packages:

In [1]:
!pip install --upgrade sqlalchemy==1.4.52

In [1]:
import pandas as pd
from sqlalchemy import create_engine

## Dataset inspection

Load the ``csv`` file in a dataframe, ``df``:

In [2]:
df = pd.read_csv('hw8_dataset_f18.csv')

In [3]:
df.head()

first_name   last_name                     email         cell_phone  \
0  Jackqueline     Dulanty       jdulanty0@goo.ne.jp  +960-657-528-2938   
1      Merrili        Fost            mfost1@nps.gov  +351-633-272-4106   
2         Shae  Haberfield      shaberfield2@ihg.com   +62-629-240-2964   
3       Kennie      Church       kchurch3@nature.com  +381-486-928-8686   
4       Shadow     Puckett  spuckett4@shareasale.com   +27-611-497-3264   

                   movie_title           movie_genres    purchase_datetime  \
0        Thirteenth Floor, The  Drama|Sci-Fi|Thriller  2018-08-27 17:29:17   
1                  Furies, The  Drama|Romance|Western  2018-08-20 08:42:38   
2                  Smart Money            Crime|Drama  2018-07-24 00:15:35   
3             Wishful Thinking                 Comedy  2018-09-22 11:01:17   
4  You Are God (Jestes Bogiem)                  Drama  2018-01-15 08:41:31   

   movie_price  
0         4.04  
1         6.96  
2        13.47  
3        13.40  
4        10.38

Always good practice to check for repeating and null values in the dataset.

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
first_name           1000 non-null object
last_name            1000 non-null object
email                1000 non-null object
cell_phone           890 non-null object
movie_title          1000 non-null object
movie_genres         1000 non-null object
purchase_datetime    1000 non-null object
movie_price          1000 non-null float64
dtypes: float64(1), object(7)
memory usage: 62.6+ KB


As we can see above, there are 1000 rows in the dataframe. All columns have been assigned "object" datatype except for movie price that was interpreted as "numeric". Also, all columns have data for 1000 rows except for "cell_phone" which has data in 890 rows, hence, 110 NULL values. We will need to take this in to account when designing the database tables. In the context of online orders, an email is perhaps more important than a phone number, hence, it should be acceptable to have NULL values for cell phone numbers.

Next, we have to check for repeating values. Are there 1000 individual movie titles? If not, is the price equal among the same movie titles? Are there 1000 individual customers?

In [5]:
df.movie_title[df.movie_title.duplicated()]

221                                            Magic Man
469                                     Strike (Stachka)
488                                          Chosen, The
498                                        Sentinel, The
524                                           Afterwards
580                                      Invisible Agent
595    Slap Her... She's French (a.k.a. She Gets What...
652    King and the Mockingbird, The (Le roi et l'ois...
694                                           Whirlygirl
727                                     Dirty Dozen, The
784                 Infernal Affairs 2 (Mou gaan dou II)
792                        I Saw the Sun (Günesi gördüm)
815                                           Whirlygirl
816                                       Borrowers, The
825                                   Secret Garden, The
841                                          Out of Time
856                                        Entitled, The
893    Only Old Men Are Going t

For movie titles, we can see that there are indeed titles which are repeating such as "Whirlygirl" and "Afterwards". What about the prices of these movies?

In [6]:
df[['movie_title', 'movie_price']][df.movie_title.isin(['Afterwards', 'Whirlygirl'])]

movie_title  movie_price
155  Whirlygirl         2.92
503  Afterwards        13.13
524  Afterwards        14.88
694  Whirlygirl         2.24
815  Whirlygirl        17.97

Movie price is different per purchase. This must serve as a hint that prices are not linked to the movies, rather, price is set at checkout. For example, a customer might have a coupon and get a discount on the movie.

Moving on to unique customers...

In [7]:
len(df[['first_name', 'last_name']].drop_duplicates())

1000

Since the length of the resulting subset of ``df`` by using ``drop_duplicates()`` is 1000 (equal to total number of rows), we can assert that within the given dataset, there are no orders made by the same person. Keep in mind however, that this might be possible in larger datasets and in any case it is reasonable to assume that one customer can purchase multiple movies.

## Normalization

- Normalization plan can be designed with a few rows of data, not the entire dataset
- Movies can have multiple genres, seperated with the pipe symbol ("|")

Let's work with the first three rows of data of the original, unnormalized dataset (ignore the index column set by Pandas):

In [8]:
df.head(3)

first_name   last_name                 email         cell_phone  \
0  Jackqueline     Dulanty   jdulanty0@goo.ne.jp  +960-657-528-2938   
1      Merrili        Fost        mfost1@nps.gov  +351-633-272-4106   
2         Shae  Haberfield  shaberfield2@ihg.com   +62-629-240-2964   

             movie_title           movie_genres    purchase_datetime  \
0  Thirteenth Floor, The  Drama|Sci-Fi|Thriller  2018-08-27 17:29:17   
1            Furies, The  Drama|Romance|Western  2018-08-20 08:42:38   
2            Smart Money            Crime|Drama  2018-07-24 00:15:35   

   movie_price  
0         4.04  
1         6.96  
2        13.47

### 1NF: Domains of all table attributes must be atomic and there cannot be repeating attributes

- Primary keys are noted with (PK)
- The original ``movie_genres`` column was not atomic and was separated into the ``genres`` table.
- The ``customers_movies`` table has a composite primary key. Rows were expanded to relate movies to genres.


#### ``customer_movies`` table:

|customer_id (PK)|genre_id (PK)|first_name|last_name|email|cell_phone|movie_title|purchase_datetime|movie_price|
|:----:|:----:|:----:|:----:|:----:|:----:|:----:|:----:|:----:|
|1|1|Jackqueline|Dulanty|jdulanty0@goo.ne.jp|+960-657-528-2938|Thirteenth Floor, The|2018-08-27 17:29:17|4.04|
|1|2|Jackqueline|Dulanty|jdulanty0@goo.ne.jp|+960-657-528-2938|Thirteenth Floor, The|2018-08-27 17:29:17|4.04|
|1|3|Jackqueline|Dulanty|jdulanty0@goo.ne.jp|+960-657-528-2938|Thirteenth Floor, The|2018-08-27 17:29:17|4.04|
|2|1|Merrili|Fost|mfost1@nps.gov|+351-633-272-4106|Furies, The|2018-08-20 08:42:38|6.96|
|2|4|Merrili|Fost|mfost1@nps.gov|+351-633-272-4106|Furies, The|2018-08-20 08:42:38|6.96|
|2|5|Merrili|Fost|mfost1@nps.gov|+351-633-272-4106|Furies, The|2018-08-20 08:42:38|6.96|
|3|6|Shae|Haberfield|shaberfield2@ihg.com|+62-629-240-2964|Smart Money|2018-07-24 00:15:35|13.47|
|3|1|Shae|Haberfield|shaberfield2@ihg.com|+62-629-240-2964|Smart Money|2018-07-24 00:15:35|13.47|
|...|...|...|...|...|...|...|...|...|



#### ``genres`` table:

|genre_id (PK)|genre|
|----|----|
|1|Drama|
|2|Sci-Fi|
|3|Thriller|
|4|Romance|
|5|Western|
|6|Crime|
|...|...|




### 2NF: Must be in 1NF and every non-key attribute must be fully dependent on the key

- ``first_name``, ``last_name``, ``email`` and ``cell_phone`` were moved to a new ``customers`` table to satisfy 2NF rules
- ``movie_title`` and ``movie_price`` were moved to a new ``movies`` table to satisfy 2NF rules.
- a new ``movie_genres`` table was created to remove the ``genre_id`` from the 1NF ``customers_movies`` table and satisfy 2NF rules. Notice that this way movies can have multiple genres and if the name of a genre is updated it will reflect on all movies.
- a new ``orders`` table was created to accomodate the ``purchase_datetime`` attribute and relate movies to customers


#### ``customer`` table:

|customer_id (PK)|first_name|last_name|email|cell_phone|
|----|----|----|----|----|
|1|Jackqueline|Dulanty|jdulanty0@goo.ne.jp|+960-657-528-2938|
|2|Merrili|Fost|mfost1@nps.gov|+351-633-272-4106|
|3|Shae|Haberfield|shaberfield2@ihg.com|+62-629-240-2964|
|...|...|...|...|...|


#### ``movies`` table:

|movie_id (PK)|movie_title|movie_price|
|----|----|----|
|1|Thirteenth Floor, The|4.04|
|2|Furies, The|6.96|
|3|Smart Money|13.47|
|...|...|...|


#### ``genres`` table:

|genre_id (PK)|genre|
|----|----|
|1|Drama|
|2|Sci-Fi|
|3|Thriller|
|4|Romance|
|5|Western|
|6|Crime|
|...|...|


#### ``movie_genres`` table:

|movie_id (PK)|genre_id (PK)|
|----|----|
|1|1|
|1|2|
|1|3|
|2|1|
|2|4|
|2|5|
|3|6|
|3|1|
|...|...|

#### ``orders`` table:

|customer_id (PK)|movie_id (PK)|purchase_datetime|
|----|----|----|
|1|1|2018-08-27 17:29:17|
|2|2|2018-08-20 08:42:38|
|3|3|2018-07-24 00:15:35|
|...|...|...|

### 3NF: Must be in 2NF and every non-key attribute must be non-transitively dependent on the key

This is a little bit tricky... When inspecting the dataset, we observed that prices for the same movie might not be the same. In which case ``movie_price`` should be part of the ``orders`` table instead of the ``movies`` table. 


#### ``customer`` table:

|customer_id (PK)|first_name|last_name|email|cell_phone|
|----|----|----|----|----|
|1|Jackqueline|Dulanty|jdulanty0@goo.ne.jp|+960-657-528-2938|
|2|Merrili|Fost|mfost1@nps.gov|+351-633-272-4106|
|3|Shae|Haberfield|shaberfield2@ihg.com|+62-629-240-2964|
|...|...|...|...|...|


#### ``movies`` table:

|movie_id (PK)|movie_title|
|----|----|
|1|Thirteenth Floor, The|
|2|Furies, The|
|3|Smart Money|
|...|...|


#### ``genres`` table:

|genre_id (PK)|genre|
|----|----|
|1|Drama|
|2|Sci-Fi|
|3|Thriller|
|4|Romance|
|5|Western|
|6|Crime|
|...|...|


#### ``movie_genres`` table:

|movie_id (PK)|genre_id (PK)|
|----|----|
|1|1|
|1|2|
|1|3|
|2|1|
|2|4|
|2|5|
|3|6|
|3|1|
|...|...|

#### ``orders`` table:

|customer_id (PK)|movie_id (PK)|movie_price|purchase_datetime|
|----|----|----|----|
|1|1|4.04|2018-08-27 17:29:17|
|2|2|6.96|2018-08-20 08:42:38|
|3|3|13.47|2018-07-24 00:15:35|
|...|...|...|...|

### Entity Relationship Diagram

During the initial design phase and normalization, it is always good to draw the Entity Relationship Diagram (ERD) in order to visualize entities and relationship sets. This often helps to determine proper cardinalities and spot mistakes.

Below is the link to the ERD as defined after normalization.

https://lucid.app/lucidchart/282a41d2-2969-451e-98f1-edb476f8e4d5/view

## Create database tables

Based on the normalization plan we are now ready to create the database tables. Before running the commands below, make sure you have created a ``movie_store`` database in pgAdmin or psql.

In [9]:
# Pass the connection string to a variable, conn_url
conn_url = 'postgresql://postgres:pwd4APAN5310@localhost/movie_store'

# Create an engine that connects to PostgreSQL server
engine = create_engine(conn_url)

# Establish a connection
connection = engine.connect()

# Pass the SQL statements that create all tables
stmt = """
    CREATE TABLE customers (
        customer_id   integer,
        first_name    varchar(50) NOT NULL,
        last_name     varchar(50) NOT NULL,
        email         varchar(50) NOT NULL,
        cell_phone    varchar(20),
        PRIMARY KEY (customer_id)
    );
    
    CREATE TABLE movies (
        movie_id      integer,
        movie_title   varchar(100) NOT NULL,
        PRIMARY KEY (movie_id)
    );
    
    CREATE TABLE genres (
        genre_id      integer,
        genre         varchar(20) NOT NULL,
        PRIMARY KEY (genre_id)
    );
    
    CREATE TABLE movie_genres (
        movie_id      integer,
        genre_id      integer,
        PRIMARY KEY (movie_id, genre_id),
        FOREIGN KEY (movie_id) REFERENCES movies (movie_id),
        FOREIGN KEY (genre_id) REFERENCES genres (genre_id)
    );
    
    CREATE TABLE orders (
        customer_id   integer,
        movie_id      integer,
        movie_price   numeric(5,2) NOT NULL,
        purchase_datetime  timestamp NOT NULL,
        PRIMARY KEY (customer_id, movie_id),
        FOREIGN KEY (customer_id) REFERENCES customers (customer_id),
        FOREIGN KEY (movie_id) REFERENCES movies (movie_id)
    );
"""

# Execute the statement to create tables
connection.execute(stmt)

## Extract, Transform and Load (ETL)

With the database and all tables created (3NF), it is now time to extract, transform and load (ETL) the dataset into the database. In order to do so we will have to perform several data transformations on the loaded dataframe, ``df`` in order to create all primary keys and maintain proper relationships. For hints, we may refer back to the normalization plan and the steps taken to produce a 3NF design.

Since all customers in the dataset are unique, we can create a new column with incrementing integer numbers for ``customer_id``.

In [10]:
df.insert(0, 'customer_id', range(1, 1 + len(df)))

Let's take a look:

In [11]:
df.head()

customer_id   first_name   last_name                     email  \
0            1  Jackqueline     Dulanty       jdulanty0@goo.ne.jp   
1            2      Merrili        Fost            mfost1@nps.gov   
2            3         Shae  Haberfield      shaberfield2@ihg.com   
3            4       Kennie      Church       kchurch3@nature.com   
4            5       Shadow     Puckett  spuckett4@shareasale.com   

          cell_phone                  movie_title           movie_genres  \
0  +960-657-528-2938        Thirteenth Floor, The  Drama|Sci-Fi|Thriller   
1  +351-633-272-4106                  Furies, The  Drama|Romance|Western   
2   +62-629-240-2964                  Smart Money            Crime|Drama   
3  +381-486-928-8686             Wishful Thinking                 Comedy   
4   +27-611-497-3264  You Are God (Jestes Bogiem)                  Drama   

     purchase_datetime  movie_price  
0  2018-08-27 17:29:17         4.04  
1  2018-08-20 08:42:38         6.96  
2  2018-07-24 00:15:35        13.47  
3  2018-09-22 11:01:17        13.40  
4  2018-01-15 08:41:31        10.38

We can now easily create a subset of ``df`` corresponding to the ``customers`` database table.

In [12]:
customers_df = df[['customer_id', 'first_name', 'last_name', 'email', 'cell_phone']]

In [13]:
customers_df.head()

customer_id   first_name   last_name                     email  \
0            1  Jackqueline     Dulanty       jdulanty0@goo.ne.jp   
1            2      Merrili        Fost            mfost1@nps.gov   
2            3         Shae  Haberfield      shaberfield2@ihg.com   
3            4       Kennie      Church       kchurch3@nature.com   
4            5       Shadow     Puckett  spuckett4@shareasale.com   

          cell_phone  
0  +960-657-528-2938  
1  +351-633-272-4106  
2   +62-629-240-2964  
3  +381-486-928-8686  
4   +27-611-497-3264

And it is now possible to load customer data to the database.

*Note: remember that the connection url, engine and connection were initialized earlier therefore we do not have to repeate them hereon.*

In [14]:
customers_df.to_sql(name='customers', con=engine, if_exists='append', index=False)

One table down, four to go...

Next, we will work on the ``movies`` table. Within the dataset there are repeating movie titles so we cannot simply add a column with incrementing integer numbers for the primary key of movies as this would lead to movies with multiple primary keys. There are several approaches to tackle this problem. One could be to write a function or a for-loop that loops through movie titles and assigns integer numbers accordingly. Since this is a relatively simple dataset, we will create a temporary dataframe with the unique movie titles only, add a column of incrementing integer numbers and then map these numbers back to the main dataframe, ``df``.

In [15]:
# Create temporary dataframe with unique movie titles
temp_movie_df = pd.DataFrame(df.movie_title.unique(), columns=['movie_title'])

# Add incrementing integers
temp_movie_df.insert(0, 'movie_id', range(1, 1 + len(temp_movie_df)))

In [16]:
temp_movie_df.head()

movie_id                  movie_title
0         1        Thirteenth Floor, The
1         2                  Furies, The
2         3                  Smart Money
3         4             Wishful Thinking
4         5  You Are God (Jestes Bogiem)

Before adding the ``movie_id`` primary key to the main dataframe, ``df``, we can push the movie data to the database:

In [17]:
temp_movie_df.to_sql(name='movies', con=engine, if_exists='append', index=False)

Now we can add a new column to the main dataframe, ``df``, for the primary key of movies, ``movie_id``. This involves two steps. First, using ``temp_movie_df`` to create a list mapping ``movie_id`` with each movie in ``df`` and then inserting this list to ``df`` as a new column.

In [18]:
# Map movie_id
movie_id_list = [temp_movie_df.movie_id[temp_movie_df.movie_title == i].values[0] for i in df.movie_title]

# Add movie_id to the main dataframe
df.insert(5, 'movie_id', movie_id_list)

Let's check that a few repeating movies have the same primary key:

In [19]:
df[['movie_id', 'movie_title', 'movie_price']][df.movie_title.isin(['Afterwards', 'Whirlygirl'])]

movie_id movie_title  movie_price
155       156  Whirlygirl         2.92
503       500  Afterwards        13.13
524       500  Afterwards        14.88
694       156  Whirlygirl         2.24
815       156  Whirlygirl        17.97

And a quick look into the expanded dataframe:

In [20]:
df.head()

customer_id   first_name   last_name                     email  \
0            1  Jackqueline     Dulanty       jdulanty0@goo.ne.jp   
1            2      Merrili        Fost            mfost1@nps.gov   
2            3         Shae  Haberfield      shaberfield2@ihg.com   
3            4       Kennie      Church       kchurch3@nature.com   
4            5       Shadow     Puckett  spuckett4@shareasale.com   

          cell_phone  movie_id                  movie_title  \
0  +960-657-528-2938         1        Thirteenth Floor, The   
1  +351-633-272-4106         2                  Furies, The   
2   +62-629-240-2964         3                  Smart Money   
3  +381-486-928-8686         4             Wishful Thinking   
4   +27-611-497-3264         5  You Are God (Jestes Bogiem)   

            movie_genres    purchase_datetime  movie_price  
0  Drama|Sci-Fi|Thriller  2018-08-27 17:29:17         4.04  
1  Drama|Romance|Western  2018-08-20 08:42:38         6.96  
2            Crime|Drama  2018-07-24 00:15:35        13.47  
3                 Comedy  2018-09-22 11:01:17        13.40  
4                  Drama  2018-01-15 08:41:31        10.38

Next, we will work with movie genres. The main task is to split multiple genres in one cell. This should be done **row-wise**, meaning, a row with multiple genres will be expanded in multiple rows, each row holding the same data for all columns except for the genre column which will hold each genre. This operation is crucial since it will create the relationship between movies and genres.

This task requires the following steps:

In [21]:
# Split genres and expand row-wise. Store in a temporary dataframe
temp_genres_df = df.movie_genres.str.split('|').apply(pd.Series, 1).stack()

# Drop index level to match main dataframe
temp_genres_df.index = temp_genres_df.index.droplevel(-1)

# Name of new column
temp_genres_df.name = 'movie_genres'

# Delete the original movie_genres columns, no longer necessary
del df['movie_genres']

# Join the original dataframe with the temporary genres 
df = df.join(temp_genres_df)

The main dataframe, ``df`` now looks like this:

In [22]:
df.head(10)

customer_id   first_name   last_name                     email  \
0            1  Jackqueline     Dulanty       jdulanty0@goo.ne.jp   
0            1  Jackqueline     Dulanty       jdulanty0@goo.ne.jp   
0            1  Jackqueline     Dulanty       jdulanty0@goo.ne.jp   
1            2      Merrili        Fost            mfost1@nps.gov   
1            2      Merrili        Fost            mfost1@nps.gov   
1            2      Merrili        Fost            mfost1@nps.gov   
2            3         Shae  Haberfield      shaberfield2@ihg.com   
2            3         Shae  Haberfield      shaberfield2@ihg.com   
3            4       Kennie      Church       kchurch3@nature.com   
4            5       Shadow     Puckett  spuckett4@shareasale.com   

          cell_phone  movie_id                  movie_title  \
0  +960-657-528-2938         1        Thirteenth Floor, The   
0  +960-657-528-2938         1        Thirteenth Floor, The   
0  +960-657-528-2938         1        Thirteenth Floor, The   
1  +351-633-272-4106         2                  Furies, The   
1  +351-633-272-4106         2                  Furies, The   
1  +351-633-272-4106         2                  Furies, The   
2   +62-629-240-2964         3                  Smart Money   
2   +62-629-240-2964         3                  Smart Money   
3  +381-486-928-8686         4             Wishful Thinking   
4   +27-611-497-3264         5  You Are God (Jestes Bogiem)   

     purchase_datetime  movie_price movie_genres  
0  2018-08-27 17:29:17         4.04        Drama  
0  2018-08-27 17:29:17         4.04       Sci-Fi  
0  2018-08-27 17:29:17         4.04     Thriller  
1  2018-08-20 08:42:38         6.96        Drama  
1  2018-08-20 08:42:38         6.96      Romance  
1  2018-08-20 08:42:38         6.96      Western  
2  2018-07-24 00:15:35        13.47        Crime  
2  2018-07-24 00:15:35        13.47        Drama  
3  2018-09-22 11:01:17        13.40       Comedy  
4  2018-01-15 08:41:31        10.38        Drama

Clearly, we now have genres that are repeating. In order to assign primary keys to the genres, we will follow the same process as with the movie titles.

In [23]:
# Create temporary dataframe with unique movie genres
temp_genre_df = pd.DataFrame(df.movie_genres.unique(), columns=['genre'])

# Add incrementing integers for genre_id
temp_genre_df.insert(0, 'genre_id', range(1, 1 + len(temp_genre_df)))

This temporary genre dataframe is:

In [24]:
temp_genre_df

genre_id               genre
0          1               Drama
1          2              Sci-Fi
2          3            Thriller
3          4             Romance
4          5             Western
5          6               Crime
6          7              Comedy
7          8              Action
8          9           Adventure
9         10           Animation
10        11  (no genres listed)
11        12              Horror
12        13             Fantasy
13        14                 War
14        15           Film-Noir
15        16             Mystery
16        17             Musical
17        18            Children
18        19         Documentary
19        20                IMAX

Before adding the ``genre_id`` primary key to the main dataframe, ``df``, we can push the genre data to the database:

In [25]:
temp_genre_df.to_sql(name='genres', con=engine, if_exists='append', index=False)

Following the same process as with movie titles, we can now add a new column to the main dataframe, ``df``, for the primary key of genres, ``genre_id``. First, using ``temp_genre_df`` to create a list mapping ``genre_id`` with each genre in ``df`` and then inserting this list to ``df`` as a new column.

In [26]:
# Map genre_id
genre_id_list = [temp_genre_df.genre_id[temp_genre_df.genre == i].values[0] for i in df.movie_genres]

# Add movie_id to the main dataframe
df.insert(9, 'genre_id', genre_id_list)

And ``df`` is now:

In [27]:
df.head(10)

customer_id   first_name   last_name                     email  \
0            1  Jackqueline     Dulanty       jdulanty0@goo.ne.jp   
0            1  Jackqueline     Dulanty       jdulanty0@goo.ne.jp   
0            1  Jackqueline     Dulanty       jdulanty0@goo.ne.jp   
1            2      Merrili        Fost            mfost1@nps.gov   
1            2      Merrili        Fost            mfost1@nps.gov   
1            2      Merrili        Fost            mfost1@nps.gov   
2            3         Shae  Haberfield      shaberfield2@ihg.com   
2            3         Shae  Haberfield      shaberfield2@ihg.com   
3            4       Kennie      Church       kchurch3@nature.com   
4            5       Shadow     Puckett  spuckett4@shareasale.com   

          cell_phone  movie_id                  movie_title  \
0  +960-657-528-2938         1        Thirteenth Floor, The   
0  +960-657-528-2938         1        Thirteenth Floor, The   
0  +960-657-528-2938         1        Thirteenth Floor, The   
1  +351-633-272-4106         2                  Furies, The   
1  +351-633-272-4106         2                  Furies, The   
1  +351-633-272-4106         2                  Furies, The   
2   +62-629-240-2964         3                  Smart Money   
2   +62-629-240-2964         3                  Smart Money   
3  +381-486-928-8686         4             Wishful Thinking   
4   +27-611-497-3264         5  You Are God (Jestes Bogiem)   

     purchase_datetime  movie_price  genre_id movie_genres  
0  2018-08-27 17:29:17         4.04         1        Drama  
0  2018-08-27 17:29:17         4.04         2       Sci-Fi  
0  2018-08-27 17:29:17         4.04         3     Thriller  
1  2018-08-20 08:42:38         6.96         1        Drama  
1  2018-08-20 08:42:38         6.96         4      Romance  
1  2018-08-20 08:42:38         6.96         5      Western  
2  2018-07-24 00:15:35        13.47         6        Crime  
2  2018-07-24 00:15:35        13.47         1        Drama  
3  2018-09-22 11:01:17        13.40         7       Comedy  
4  2018-01-15 08:41:31        10.38         1        Drama

We do not need to assign any other primary keys. For the remaining two tables in our schema, we must only slice the main dataframe, eliminate duplicate rows and push to the database.

Let's focus on ``movie_genres``.

In [28]:
df[['movie_id', 'genre_id']].head(20)

movie_id  genre_id
0         1         1
0         1         2
0         1         3
1         2         1
1         2         4
1         2         5
2         3         6
2         3         1
3         4         7
4         5         1
5         6         1
6         7         8
6         7         6
6         7         1
6         7         3
7         8         8
7         8         9
7         8         5
8         9         3
9        10         1

In order to check if there are any duplicate rows we should look into the rows for repeating movies, for example, "Magic Man".

In [29]:
df[df.movie_title == 'Magic Man']

customer_id first_name   last_name                        email  \
192          193     Johnny  Fullerlove  jfullerlove5c@economist.com   
192          193     Johnny  Fullerlove  jfullerlove5c@economist.com   
192          193     Johnny  Fullerlove  jfullerlove5c@economist.com   
221          222     Lauree    Dingwall     ldingwall65@springer.com   
221          222     Lauree    Dingwall     ldingwall65@springer.com   
221          222     Lauree    Dingwall     ldingwall65@springer.com   

            cell_phone  movie_id movie_title    purchase_datetime  \
192  +351-495-428-5920       193   Magic Man  2018-06-27 16:56:51   
192  +351-495-428-5920       193   Magic Man  2018-06-27 16:56:51   
192  +351-495-428-5920       193   Magic Man  2018-06-27 16:56:51   
221  +212-807-564-8447       193   Magic Man  2017-06-09 07:07:50   
221  +212-807-564-8447       193   Magic Man  2017-06-09 07:07:50   
221  +212-807-564-8447       193   Magic Man  2017-06-09 07:07:50   

     movie_price  genre_id movie_genres  
192        14.77         6        Crime  
192        14.77        16      Mystery  
192        14.77         3     Thriller  
221        13.73         6        Crime  
221        13.73        16      Mystery  
221        13.73         3     Thriller

For table ``movie_genres``, the composite primary key is made up from ``movie_id`` and ``genre_id``. As is made obvious above, pushing this ``df`` slide to the database without removinf duplicates would return an error since both Johny and Lauree purchased "Magic Man" (movie_id = 193 with gernre_id 6, 16, 3).

Dropping duplicate rows and pushing to the database can be done in one command:

In [30]:
df[['movie_id', 'genre_id']]. \
    drop_duplicates().to_sql(name='movie_genres', con=engine, if_exists='append', index=False)

Finally, the process is the same for the ``orders`` table. If we look at the slice:

In [31]:
df[['customer_id', 'movie_id', 'movie_price', 'purchase_datetime']].head()

customer_id  movie_id  movie_price    purchase_datetime
0            1         1         4.04  2018-08-27 17:29:17
0            1         1         4.04  2018-08-27 17:29:17
0            1         1         4.04  2018-08-27 17:29:17
1            2         2         6.96  2018-08-20 08:42:38
1            2         2         6.96  2018-08-20 08:42:38

There are clearly duplicate rows that need to be removed before pushing to the database.

In [32]:
df[['customer_id', 'movie_id', 'movie_price', 'purchase_datetime']].drop_duplicates().head()

customer_id  movie_id  movie_price    purchase_datetime
0            1         1         4.04  2018-08-27 17:29:17
1            2         2         6.96  2018-08-20 08:42:38
2            3         3        13.47  2018-07-24 00:15:35
3            4         4        13.40  2018-09-22 11:01:17
4            5         5        10.38  2018-01-15 08:41:31

Which we can also do in one command:

In [33]:
df[['customer_id', 'movie_id', 'movie_price', 'purchase_datetime']]. \
    drop_duplicates().to_sql(name='orders', con=engine, if_exists='append', index=False)

## Spot Checks & Validation

With all database tables populated, we must run some tests to make sure that data was tranferred properly and all relations have been upheld.

Let's take a quick look at the original dataset:

In [34]:
df2 = pd.read_csv('hw8_dataset_f18.csv')
df2.head()

first_name   last_name                     email         cell_phone  \
0  Jackqueline     Dulanty       jdulanty0@goo.ne.jp  +960-657-528-2938   
1      Merrili        Fost            mfost1@nps.gov  +351-633-272-4106   
2         Shae  Haberfield      shaberfield2@ihg.com   +62-629-240-2964   
3       Kennie      Church       kchurch3@nature.com  +381-486-928-8686   
4       Shadow     Puckett  spuckett4@shareasale.com   +27-611-497-3264   

                   movie_title           movie_genres    purchase_datetime  \
0        Thirteenth Floor, The  Drama|Sci-Fi|Thriller  2018-08-27 17:29:17   
1                  Furies, The  Drama|Romance|Western  2018-08-20 08:42:38   
2                  Smart Money            Crime|Drama  2018-07-24 00:15:35   
3             Wishful Thinking                 Comedy  2018-09-22 11:01:17   
4  You Are God (Jestes Bogiem)                  Drama  2018-01-15 08:41:31   

   movie_price  
0         4.04  
1         6.96  
2        13.47  
3        13.40  
4        10.38

We will query for, say, the genres of the movie "The Furies".

*Note: remember that the connection url, engine and connection were initialized earlier therefore we do not have to repeate them hereon.*

In [35]:
# Pass the SQL statement to filter data
stmt = """

    SELECT g.genre 
    FROM genres g
    JOIN movie_genres mg ON g.genre_id = mg.genre_id
    JOIN movies m ON m.movie_id = mg.movie_id
    WHERE m.movie_title = 'Furies, The';

"""

# Execute the statement and get the results
results = connection.execute(stmt).fetchall()

# Extract column names
column_names = results[0].keys()

# Store results in a new dataframe
temp_df = pd.DataFrame(results, columns=column_names)

# Show results
temp_df

genre
0    Drama
1  Romance
2  Western

So far so good. It would be wise to test the same query on movie titles that were repeating, such as "Magic Man" to make sure that genres for this movie have not been stored more than once.

In [36]:
df2[df2.movie_title == 'Magic Man']

first_name   last_name                        email         cell_phone  \
192     Johnny  Fullerlove  jfullerlove5c@economist.com  +351-495-428-5920   
221     Lauree    Dingwall     ldingwall65@springer.com  +212-807-564-8447   

    movie_title            movie_genres    purchase_datetime  movie_price  
192   Magic Man  Crime|Mystery|Thriller  2018-06-27 16:56:51        14.77  
221   Magic Man  Crime|Mystery|Thriller  2017-06-09 07:07:50        13.73

In [37]:
# Pass the SQL statement to filter data
stmt = """

    SELECT g.genre 
    FROM genres g
    JOIN movie_genres mg ON g.genre_id = mg.genre_id
    JOIN movies m ON m.movie_id = mg.movie_id
    WHERE m.movie_title = 'Magic Man';

"""

# Execute the statement and get the results
results = connection.execute(stmt).fetchall()

# Extract column names
column_names = results[0].keys()

# Store results in a new dataframe
temp_df = pd.DataFrame(results, columns=column_names)

# Show results
temp_df

genre
0     Crime
1   Mystery
2  Thriller

Great! Let's get a quick count of all movies that have been purchased more than once:

In [38]:
# Pass the SQL statement to filter data
stmt = """

    SELECT m.movie_title, COUNT(m.movie_title) AS movie_count
    FROM movies m
    JOIN orders o ON o.movie_id = m.movie_id
    GROUP BY m.movie_title
    HAVING COUNT(m.movie_title) > 1
    ORDER BY movie_count DESC;

"""

# Execute the statement and get the results
results = connection.execute(stmt).fetchall()

# Extract column names
column_names = results[0].keys()

# Store results in a new dataframe
temp_df = pd.DataFrame(results, columns=column_names)

# Show results
temp_df

movie_title  movie_count
0                                          Whirlygirl            3
1   Only Old Men Are Going to Battle (V boy idut o...            2
2                                         The Beehive            2
3                                     Invisible Agent            2
4                                    Strike (Stachka)            2
5                       I Saw the Sun (Günesi gördüm)            2
6                                          Afterwards            2
7                                      Borrowers, The            2
8                                  Burn After Reading            2
9   Slap Her... She's French (a.k.a. She Gets What...            2
10                                   Dirty Dozen, The            2
11               Infernal Affairs 2 (Mou gaan dou II)            2
12                                      Entitled, The            2
13                                        Chosen, The            2
14                                        Out of Time            2
15  King and the Mockingbird, The (Le roi et l'ois...            2
16                                      Sentinel, The            2
17                                 Secret Garden, The            2
18                                          Magic Man            2

It appears that the most popular movie was "Whirlygirl". Similarly, we can get counts of movie genres:

In [39]:
# Pass the SQL statement to filter data
stmt = """

    SELECT g.genre, COUNT(g.genre) AS genre_count
    FROM genres g
    JOIN movie_genres mg ON g.genre_id = mg.genre_id
    GROUP BY g.genre
    ORDER BY genre_count DESC;

"""

# Execute the statement and get the results
results = connection.execute(stmt).fetchall()

# Extract column names
column_names = results[0].keys()

# Store results in a new dataframe
temp_df = pd.DataFrame(results, columns=column_names)

# Show results
temp_df

genre  genre_count
0                Drama          479
1               Comedy          293
2             Thriller          167
3              Romance          131
4               Action          125
5                Crime          104
6               Horror          103
7            Adventure           81
8          Documentary           79
9              Mystery           55
10              Sci-Fi           53
11             Fantasy           52
12                 War           47
13            Children           38
14             Musical           37
15             Western           33
16           Animation           30
17           Film-Noir           15
18  (no genres listed)           10
19                IMAX            7

Finally, since "Whirlygirl" was sold more than once, let's look at when it was purchased and by whom:

In [40]:
# Pass the SQL statement to filter data
stmt = """

    SELECT c.first_name, c.last_name, m. movie_title, o.movie_price, o.purchase_datetime
    FROM orders o
    JOIN movies m ON m.movie_id = o.movie_id
    JOIN customers c ON c.customer_id = o.customer_id
    WHERE m.movie_title = 'Whirlygirl';

"""

# Execute the statement and get the results
results = connection.execute(stmt).fetchall()

# Extract column names
column_names = results[0].keys()

# Store results in a new dataframe
temp_df = pd.DataFrame(results, columns=column_names)

# Show results
temp_df

first_name last_name movie_title movie_price   purchase_datetime
0      Madel  Kirkland  Whirlygirl        2.92 2017-06-04 08:21:41
1      Elwyn   Furness  Whirlygirl        2.24 2017-02-05 21:26:18
2       Eyde   Lippitt  Whirlygirl       17.97 2018-04-29 07:37:39

The result matching the original dataset for the same movie:

In [41]:
df2[df2.movie_title == 'Whirlygirl']

first_name last_name                        email        cell_phone  \
155      Madel  Kirkland  mkirkland4b@arstechnica.com  +63-712-138-5062   
694      Elwyn   Furness          efurnessja@xrea.com  +86-618-650-7587   
815       Eyde   Lippitt      elippittmn@virginia.edu  +86-120-329-8613   

    movie_title   movie_genres    purchase_datetime  movie_price  
155  Whirlygirl  Drama|Romance  2017-06-04 08:21:41         2.92  
694  Whirlygirl  Drama|Romance  2017-02-05 21:26:18         2.24  
815  Whirlygirl  Drama|Romance  2018-04-29 07:37:39        17.97

## Conclusion

This exercise was a basic normalization and ETL process aimed at building your skills in developing and populating relational databases. A small change in the original dataset could lead to a completely different database design and ETL process. It is therefore very important to develop your coding skills so that you are comfortable and flexible implementing different approaches. There is no effective book or guide on how to implement ETL processes. It all comes down to creativity and lots of practice.